### PR TITLE
SANY commandline enhancements

### LINKED ISSUE #947: SANY should output command line help text
Related to #783.

SANY should output help text for its command line parameters in the following cases:

1. No parameters are given
2. Incorrect or unrecognized parameters are given

Currently the only way to figure out how the SANY command line parameters work is to read the code that processes them.

There is a `-help` parameter, but the documentation it prints out is incomplete. Also, it does not document the debug spec explorer shell commands.

### PR DESCRIPTION
**Pull request by MathieuBordere**
Fixes #947 

- Show usage when no argument is provided.
- Show error and refer to `-help` when no file argument is provided.
- Slight rewording of existing options.
- Added mandatory `FILE...` argument.
- Added explanation for commands that can be used in the Semantic Graph Explorer.

This results in:

```
NAME

	SANY - provides parsing, semantic analysis, and level-checking for a TLA+ spec - Version 2.2 created 08 July 2020


SYNOPSIS

	SANY [-dls] [-error-codes] FILE...

DESCRIPTION

	SANY is a parser and syntax checker for TLA+ specifications.
	It catches parsing errors and some "semantic" errors such as
	priming an expression containing primed variables.

OPTIONS

	-d
		Opens the semantic graph explorer prompt. The prompt accepts the following commands:
		'cst' Prints out the concrete syntax tree.
		'dot' Emits the semantic graph to a ModuleName.dot file in the DOT graph description language.
		'mt'  Prints the module table, a list of all imported modules and their top-level definitions.
		'mt*' Prints the extended module table including built-in operator definitions.
		
		Optionally, skip the prompt and run a command directly by appending it like 'SANY -d ModuleName.tla cst'
	-error-codes
		Return a descriptive exit code in case of error.
		
		'1' Error during initialization.
		'2' Error during parsing.
		'3' Error parsing built-in operator tables.
		'4' Error during semantic analysis or level-checking
	-l
		Turns off level-checking.
	-s
		Turns off semantic analysis and level-checking.
```

### CODE REVIEW

File: tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java [hunk 1]
```diff
@@ -408,6 +422,11 @@ private static void printUsage()
    * ExternalModuleTable.
    */
   public static void SANYmain(String args[]) {
+    if (args.length == 0) {
+      printUsage();
+      System.exit(-1);
```
Comment (ahelwer) [inline tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:427]: In keeping with the (admittedly, bad) precedent of always returning a zero exit code unless `-error-codes` is given, perhaps we should just do a `return` here to exit with 0.

File: tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java [hunk 2]
```diff
@@ -430,12 +449,18 @@ else if (args[i].toLowerCase().equals("-help")) {
...
+           ToolIO.out.println("Invalid option: " + args[i]);
+           ToolIO.out.println("Try 'SANY -help' for more information.");
            System.exit(-1);
       }
     }
 
+    if (i == args.length) {
+      ToolIO.out.println("At least 1 filename is required.");
+      ToolIO.out.println("Try 'SANY -help' for more information.");
+      System.exit(-1);
```
Comment (ahelwer) [inline tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:461]: Same as above

File: tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java [hunk 3]
```diff
@@ -430,12 +449,18 @@ else if (args[i].toLowerCase().equals("-help")) {
            printUsage();
            System.exit(0);
       } else {
-           ToolIO.out.println("Command-line error: " + args[i]);
-           ToolIO.out.println("Use -help option for more information.");
+           ToolIO.out.println("Invalid option: " + args[i]);
+           ToolIO.out.println("Try 'SANY -help' for more information.");
            System.exit(-1);
```
Comment (ahelwer) [inline tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:454]: Same as above

Comment (ahelwer) [reply tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:454]: Ah, I guess this was already present. Well then we have the precedent of returning -1 on bad parameters so keep all of these in.

File: tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java [hunk 4]
```diff
@@ -382,13 +382,27 @@ private static void printUsage()
...
-          "used, if the semantic analysis is disabled. ", true));
+          "-l", "Turns off level-checking.", true));
       variant.add(new UsageGenerator.Argument(
-          "-error-codes", "If enabled, error level will be reported as the tools'\n" +
-          "return value." , true));
+          "-error-codes", "Return a descriptive exit code in case of error.\n\n" +
+          "'1' Error during initialization.\n" +
+          "'2' Error during parsing.\n" +
+          "'3' Error parsing built-in operator tables.\n" +
+          "'4' Error during semantic analysis or level-checking.", true));
```
Comment (ahelwer) [inline tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:393]: I am pretty sure the possibility of an error occurring during either initialization or parsing built-in operator tables (or both) was removed by #1012. So we might not want to document these.

Comment (ahelwer) [reply tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:393]: Yeah I confirmed it's impossible for these errors to ever occur, although the error reporting machinery still exists. I took out this machinery in a new PR #1069.

Comment (MathieuBordere) [reply tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java]: I've removed the 2 cases.

**Review by ahelwer — APPROVED** [review]

File: tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java [hunk 5]
```diff
@@ -430,12 +447,18 @@ else if (args[i].toLowerCase().equals("-help")) {
            printUsage();
            System.exit(0);
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:448]: How about consistently existing with `System.exit(0)` here and above if no arguments are given?

Comment (MathieuBordere) [reply tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:448]: The shell tools I've tried that accept mandatory file arguments like `file`, `grep`, `chmod`, `rm`, `mv`, `cp` all exit with non-0 exit codes when they're given no arguments, I tried to be consistent with that. If you would still prefer to exit with `System.exit(0)` I can change it.

**Review by lemmy — APPROVED** [review]

File: tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java (hunk 5 above)
Comment (MathieuBordere) [reply tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:448]: Hmm, I'd rather put my time in something less administrative :-), let me know if you want me to change some exit values in this PR still.

Comment (lemmy) [reply tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:448]: Fine by me as long as commits are focused on small, specific changes.

Comment (MathieuBordere) [reply tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:448]: I think you misunderstood what I meant, I'd rather not to this in this PR, please let me know if I need to change exit values of SANY somewhere to get this PR merged.

Comment (lemmy) [reply tlatools/org.lamport.tlatools/src/tla2sany/drivers/SANY.java:448]: No :-)

**Pull request merged**